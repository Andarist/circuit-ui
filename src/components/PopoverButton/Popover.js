import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Manager, Reference, Popper } from 'react-popper';
import styled from 'react-emotion';

const modifiers = {
  offset: { enabled: true, offset: '0,10' },
  flip: { enabled: true }
};

class Popover extends Component {
  static propTypes = {
    /**
     * function rendering the popover
     */
    renderPopover: PropTypes.func.isRequired,
    /**
     * function rendering the reference (button or something clickable)
     */
    renderReference: PropTypes.func.isRequired,
    /**
     * placement of the popover relative to the button
     */
    placement: PropTypes.oneOf(['top', 'bottom', 'left', 'right']),
    /**
     * alignment of the popover relative to the button
     */
    align: PropTypes.oneOf(['start', 'end', 'center'])
  };

  static defaultProps = {
    placement: 'bottom',
    align: 'start'
  };

  state = { isOpen: false };

  componentDidMount() {
    document.addEventListener('click', this.onDocumentClick, false);
  }

  componentWillUnmount() {
    document.removeEventListener('click', this.onDocumentClick, false);
  }

  onDocumentClick = ({ target }) => {
    if (this.popoverRef && this.popoverRef.contains(target)) {
      return;
    }
    if (this.buttonRef && this.buttonRef.contains(target)) {
      return;
    }

    this.handleOutsideClick();
  };

  buttonRef = null;
  popoverRef = null;

  receiveButtonRef = ref => {
    this.buttonRef = ref;
  };

  receivePopoverRef = ref => {
    this.popoverRef = ref;
  };

  handleClick = () => this.setState(({ isOpen }) => ({ isOpen: !isOpen }));
  handleOutsideClick = () => {
    this.setState({ isOpen: false });
  };

  render() {
    const { renderPopover, renderReference, placement, align } = this.props;
    const { isOpen } = this.state;

    return (
      <Manager>
        <Reference>
          {({ ref }) => (
            <ButtonWrap
              innerRef={this.receiveButtonRef}
              onClick={this.handleClick}
            >
              <div ref={ref}>{renderReference({ isOpen })}</div>
            </ButtonWrap>
          )}
        </Reference>
        <Popper
          placement={toPopperPlacement(placement, align)}
          modifiers={modifiers}
        >
          {({ ref, style }) =>
            isOpen && (
              <div ref={this.receivePopoverRef}>
                <div {...{ ref, style }}>{renderPopover()}</div>
              </div>
            )
          }
        </Popper>
      </Manager>
    );
  }
}

function toPopperPlacement(placement, align) {
  return placement + (align !== 'center' ? `-${align}` : '');
}

const ButtonWrap = styled('div')`
  display: inline-block;
`;

/**
 * @component
 */
export default Popover;
