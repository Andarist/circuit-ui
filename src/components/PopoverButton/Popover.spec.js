import React from 'react';

import Popover from '.';

// FMI: https://github.com/FezVrasta/popper.js/issues/478
jest.mock('popper.js', () => {
  const PopperJS = jest.requireActual('popper.js');

  return class Popper {
    static placements = PopperJS.placements;

    constructor() {
      return {
        destroy: () => {},
        scheduleUpdate: () => {}
      };
    }
  };
});

describe('Popover', () => {
  beforeEach(() => {
    jest.resetAllMocks();
  });

  /**
   * Style tests.
   */
  it('should render with default styles', () => {
    const actual = create(
      <Popover renderReference={() => <span />} renderPopover={() => <div />} />
    );
    expect(actual).toMatchSnapshot();
  });

  /**
   * Accessibility tests.
   */
  it('should meet accessibility guidelines', async () => {
    const wrapper = renderToHtml(
      <Popover renderReference={() => <span />} renderPopover={() => <div />} />
    );
    const actual = await axe(wrapper);
    expect(actual).toHaveNoViolations();
  });

  const placements = ['top', 'right', 'bottom', 'left'];
  const alignments = ['start', 'end', 'center'];

  placements.forEach(placement => {
    alignments.forEach(alignment => {
      it(`should render with placement ${placement} and alignment ${alignment}`, () => {
        const actual = create(
          <Popover
            placement={placement}
            align={alignment}
            renderReference={() => <span />}
            renderPopover={() => <span />}
          />
        );
        expect(actual).toMatchSnapshot();
      });
    });
  });

  it('calls renderPopover on button click', () => {
    const renderFn = jest.fn();
    const actual = mount(
      <Popover
        renderReference={() => <button>Button</button>}
        renderPopover={renderFn}
      />
    );
    const button = actual.find('button');

    expect(renderFn).toHaveBeenCalledTimes(0);
    button.simulate('click');
    expect(renderFn).toHaveBeenCalled();
  });

  it('calls renderReference with isOpen', () => {
    const renderFn = jest.fn(() => <button>button</button>);
    const actual = mount(
      <Popover renderReference={renderFn} renderPopover={() => <span />} />
    );
    const button = actual.find('button');
    button.simulate('click');

    expect(renderFn).toHaveBeenCalledWith({ isOpen: true });
  });
});
