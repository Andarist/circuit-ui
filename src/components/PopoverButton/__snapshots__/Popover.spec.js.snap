// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popover should render with default styles 1`] = `
.circuit-0 {
  display: inline-block;
}

<div
  className="circuit-0 circuit-1"
  onClick={[Function]}
>
  <div>
    <span />
  </div>
</div>
`;

exports[`Popover should render with placement bottom and alignment center 1`] = `
.circuit-0 {
  display: inline-block;
}

<div
  className="circuit-0 circuit-1"
  onClick={[Function]}
>
  <div>
    <span />
  </div>
</div>
`;

exports[`Popover should render with placement bottom and alignment end 1`] = `
.circuit-0 {
  display: inline-block;
}

<div
  className="circuit-0 circuit-1"
  onClick={[Function]}
>
  <div>
    <span />
  </div>
</div>
`;

exports[`Popover should render with placement bottom and alignment start 1`] = `
.circuit-0 {
  display: inline-block;
}

<div
  className="circuit-0 circuit-1"
  onClick={[Function]}
>
  <div>
    <span />
  </div>
</div>
`;

exports[`Popover should render with placement left and alignment center 1`] = `
.circuit-0 {
  display: inline-block;
}

<div
  className="circuit-0 circuit-1"
  onClick={[Function]}
>
  <div>
    <span />
  </div>
</div>
`;

exports[`Popover should render with placement left and alignment end 1`] = `
.circuit-0 {
  display: inline-block;
}

<div
  className="circuit-0 circuit-1"
  onClick={[Function]}
>
  <div>
    <span />
  </div>
</div>
`;

exports[`Popover should render with placement left and alignment start 1`] = `
.circuit-0 {
  display: inline-block;
}

<div
  className="circuit-0 circuit-1"
  onClick={[Function]}
>
  <div>
    <span />
  </div>
</div>
`;

exports[`Popover should render with placement right and alignment center 1`] = `
.circuit-0 {
  display: inline-block;
}

<div
  className="circuit-0 circuit-1"
  onClick={[Function]}
>
  <div>
    <span />
  </div>
</div>
`;

exports[`Popover should render with placement right and alignment end 1`] = `
.circuit-0 {
  display: inline-block;
}

<div
  className="circuit-0 circuit-1"
  onClick={[Function]}
>
  <div>
    <span />
  </div>
</div>
`;

exports[`Popover should render with placement right and alignment start 1`] = `
.circuit-0 {
  display: inline-block;
}

<div
  className="circuit-0 circuit-1"
  onClick={[Function]}
>
  <div>
    <span />
  </div>
</div>
`;

exports[`Popover should render with placement top and alignment center 1`] = `
.circuit-0 {
  display: inline-block;
}

<div
  className="circuit-0 circuit-1"
  onClick={[Function]}
>
  <div>
    <span />
  </div>
</div>
`;

exports[`Popover should render with placement top and alignment end 1`] = `
.circuit-0 {
  display: inline-block;
}

<div
  className="circuit-0 circuit-1"
  onClick={[Function]}
>
  <div>
    <span />
  </div>
</div>
`;

exports[`Popover should render with placement top and alignment start 1`] = `
.circuit-0 {
  display: inline-block;
}

<div
  className="circuit-0 circuit-1"
  onClick={[Function]}
>
  <div>
    <span />
  </div>
</div>
`;
